### PR TITLE
Chunking on cold

### PR DESCRIPTION
**Pull request by zachm**
Still todo: write tests. I thought y'all might want this with your morning coffee, though :)

I use pull/6 to split the time range into chunks, then use naive Python threading to parallelize the KairosDB calls for those chunks.

There's a bit of Magic in KQuery.from_request which allows us to perfectly merge chunks together. Otherwise, it's complex but not very complicated?

Will update this PR tomorrow with some test coverage.

### CODE REVIEW

File: tscached/kquery.py [hunk 1]
```diff
@@ -22,6 +24,16 @@ def from_request(cls, request, redis_client):
         for metric in request.get('metrics', []):
             new = cls(redis_client)
             new.query = metric
+
+            # This seemingly terrifying kludge makes it easy to merge chunks of MTS together.
+            # For all aggregators, we remove align_sampling and add align_start_time.
+            # Otherwise, partial windows (if aggregated) will appear on every seam in the data. Info:
+            # https://kairosdb.github.io/docs/build/html/restapi/QueryMetrics.html#metric-properties
+            for agg_ndx in xrange(len(new.query.get('aggregators', []))):
```
Comment (tiras-j) [inline tscached/kquery.py:32]: Do we handle any aggregators in KQuery's currently? Is grafana (or whatever) able to request a data set with an aggregator enabled? 

Also, on a side note, would we want to implement aggregators for large (i.e. 3 hour) data sets as an effective form of downsampling to reduce the total data returned and maybe speed up query results? j/w

File: tscached/kquery.py [hunk 2]
```diff
@@ -60,8 +72,51 @@ def proxy_to_kairos(self, host, port, time_range):
...
+        for time_range in time_ranges:
+            # Build a full thread, with query, out of each chunk of time.
+            start_ts = int(time_range[0].strftime('%s')) * 1000
+            end_ts = int(time_range[1].strftime('%s')) * 1000
+            query = {'start_absolute': start_ts, 'end_absolute': end_ts, 'cache_time': 0}
+            query['metrics'] = [self.query]
+
+            # Create the thread, keep a reference to it, and start it off.
+            thr = threading.Thread(target=_thread_wrap, args=(ndx, query))
+            thr.daemon = True
```
Comment (tiras-j) [inline tscached/kquery.py]: If we're joining the thread we probably shouldn't daemonize it.

File: tscached.yaml [hunk 3]
```diff
@@ -27,8 +27,9 @@ tscached:
         staleness_threshold: 10  # data up to this far in the past is "new"
 
     chunking:
-        chunk_length: 1800  # chunk on 30 minute intervals
+        chunk_length: 3600  # chunk on 1 hour intervals
```
Comment (tiras-j) [inline tscached.yaml:30]: Do we want to update the default in "get_chunked_time_ranges"? Currently it defaults to 1800 if no config is defined.

File: tscached/kquery.py [hunk 4]
```diff
@@ -22,6 +24,16 @@ def from_request(cls, request, redis_client):
...
             new = cls(redis_client)
             new.query = metric
+
+            # This seemingly terrifying kludge makes it easy to merge chunks of MTS together.
+            # For all aggregators, we remove align_sampling and add align_start_time.
+            # Otherwise, partial windows (if aggregated) will appear on every seam in the data. Info:
+            # https://kairosdb.github.io/docs/build/html/restapi/QueryMetrics.html#metric-properties
+            for agg_ndx in xrange(len(new.query.get('aggregators', []))):
+                if 'align_sampling' in new.query['aggregators'][agg_ndx]:
+                    del new.query['aggregators'][agg_ndx]['align_sampling']
```
Comment (tiras-j) [inline tscached/kquery.py:34]: Also do we want to delete this reference or simply set it to False? Since it appears to be a boolean value.

File: tscached.yaml [hunk 5]
```diff
@@ -27,8 +27,9 @@ tscached:
...
         max_chunks: 6  # increase chunk size if more than this needed
+        thread_timeout: 30  # timeout on waiting for chunk threads to join
```
Comment (venkey-ariv) [inline tscached.yaml:32]: We might want to set this to Kairos query timeout + some small duration. 

With chunking of 1 hr, we might still have a very high volume query that returns in > 30 seconds.

**Comment by venkey-ariv** [comment]
Fix and ship! :)

File: tscached/kquery.py (hunk 1 above)
Comment (zachm) [inline tscached/kquery.py:32]: To the first point, yeah. This is sum, diff, sample, etc.

To the second, oh hell yeah. It's probably scope creep for now. My initial intuition was to have tscached be as naive as possible to what the user wants. Just take their word for it as best we can?

File: tscached/kquery.py (hunk 4 above)
Comment (zachm) [inline tscached/kquery.py:34]: Just following the KairosDB spec - they claim that setting both values leads to unpredictable behavior. Same difference, I figure.

File: tscached.yaml (hunk 5 above)
Comment (zachm) [inline tscached.yaml:32]: I actually don't want to do this. If you use the raw KairosDB internally, it has a very high timeout: 1 minute if I recall correctly. We'd rather time out before that in the thread. 1 minute is necessary for long (over 3 hours) queries. It's configurable, so we can always make a different choice later :)
